libsolidity/analysis/PostTypeContractLevelChecker.cpp

```diff
@@ -145,3 +147,43 @@ void PostTypeContractLevelChecker::checkStorageLayoutSpecifier(ContractDefinitio
 			"Contract extends past the end of storage when this base slot value is specified."
 		);
 }
+
+void PostTypeContractLevelChecker::warnStorageLayoutBaseNearStorageEnd(ContractDefinition const& _contract)
+{
+	// In case of most errors the warning is pointless. E.g. if we're already past storage end.
+	// If the errors were in the layout specifier, we may not even be able to get values to validate.
+	if (Error::containsErrors(m_errorReporter.errors()))
+		return;
+
+	bigint storageSize = contractStorageSizeUpperBound(_contract, VariableDeclaration::Location::Unspecified);
+	u256 baseSlot = layoutBaseForInheritanceHierarchy(_contract, DataLocation::Storage);
+	solAssert(baseSlot + storageSize <= std::numeric_limits<u256>::max());
+
+	if (
+		u256 slotsLeft = std::numeric_limits<u256>::max() - baseSlot - u256(storageSize);
+		slotsLeft <= u256(1) << 64
+	)
+	{
+		auto const& location = _contract.storageLayoutSpecifier() ?
+			_contract.storageLayoutSpecifier()->location() :
+			_contract.location();
+
+		auto errorID = 3495_error;
+		std::string errorMsg = "This contract is very close to the end of storage. This limits its future upgradability.";
+		if (_contract.stateVariables().size() > 0)
```

Comment:
Wait, no. This is also wrong. It includes all state variables rather than just ones in storage.